src/terms.jl

```diff
@@ -398,18 +376,52 @@ Base.show(io::IO, mime::MIME"text/plain", t::MatrixTerm; prefix="") =
 # operators on Terms that create new terms:
 
 
-Base.:~(lhs::TermOrTerms, rhs::TermOrTerms) = FormulaTerm(lhs, rhs)
+Base.:~(lhs::TermOrTerms, rhs::TermOrTerms) = FormulaTerm(lhs, cleanup(rhs))
 
-Base.:&(terms::AbstractTerm...) = InteractionTerm(terms)
 Base.:&(term::AbstractTerm) = term
-Base.:&(it::InteractionTerm, terms::AbstractTerm...) = InteractionTerm((it.terms..., terms...))
-
+Base.:&(a::AbstractTerm, b::AbstractTerm) = InteractionTerm((a,b))
+
+Base.:&(::ConstantTerm, b::AbstractTerm) = b
+Base.:&(a::AbstractTerm, ::ConstantTerm) = a
+
+# Avoid method ambiguities
+Base.:&(::ConstantTerm, b::InteractionTerm) = b
+Base.:&(a::InteractionTerm, ::ConstantTerm) = a
+Base.:&(a::ConstantTerm, ::ConstantTerm) = a
+
+# associative rule
+Base.:&(it::InteractionTerm, term::AbstractTerm) =
+    term in it.terms ? it : InteractionTerm((it.terms..., term))
+Base.:&(term::AbstractTerm, it::InteractionTerm) =
+    term in it.terms ? it : InteractionTerm((term, it.terms...))
+Base.:&(a::InteractionTerm, b::InteractionTerm) =
+    InteractionTerm((union(a.terms, b.terms)..., ))
+
+# distributive rule
+Base.:&(term::AbstractTerm, terms::TupleTerm) = term .& terms
+Base.:&(terms::TupleTerm, term::AbstractTerm) = terms .& term
+Base.:&(as::TupleTerm, bs::TupleTerm) = ((a & b for a in as for b in bs)..., )
+
+# + concatenates terms
 Base.:+(a::AbstractTerm) = a
 Base.:+(a::AbstractTerm, b::AbstractTerm) = a==b ? a : (a, b)
+
+# associative rule for +
 Base.:+(as::TupleTerm, b::AbstractTerm) = b in as ? as : (as..., b)
 Base.:+(a::AbstractTerm, bs::TupleTerm) = a in bs ? bs : (a, bs...)
 Base.:+(as::TupleTerm, bs::TupleTerm) = (union(as, bs)..., )
 
+# * expansion
+Base.:*(a::TermOrTerms, b::TermOrTerms) = a + b + a&b
+
+cleanup(terms::TupleTerm) = tuple(sort!(unique!([terms...]), by=degree)...)
+cleanup(x) = x
+
+degree(::AbstractTerm) = 1
+degree(t::InteractionTerm) = mapreduce(degree, +, t.terms)
+# dirty hack, move to MixedModels.jl
+degree(::FunctionTerm{typeof(|)}) = Inf
```

Comment:
This means that this PR will actually be breaking if any packages define similar terms?

Reply:
"similar terms" meaning using `typeof(|)` for special syntax?

Reply:
No, since apparently this PR requires changing something in MixedModels or adding a hack here, could it affect other packages too?

Reply:
ah yeah I see now.  we are already special casing calls to `:|` so this is not breaking:

https://github.com/JuliaStats/StatsModels.jl/blob/463eb0acb49bc5428374d749c4da90ea2a6c74c4/src/formula.jl#L272-L274